stl/inc/flat_map

```diff
@@ -0,0 +1,1003 @@
+// flat_map standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef _FLAT_MAP_
+#define _FLAT_MAP_
+#include <yvals_core.h>
+#if _STL_COMPILER_PREPROCESSOR
+#if !_HAS_CXX23 || !defined(__cpp_lib_concepts) // TRANSITION, GH-395
+_EMIT_STL_WARNING(STL4038, "The contents of <flat_map> are available only with C++23 or later.");
+#else // ^^^ not supported / supported language mode vvv
+
+#include <algorithm>
+#include <compare>
+#include <concepts>
+#include <cstdint>
+#include <initializer_list>
+#include <memory>
+#include <ranges>
+#include <type_traits>
+#include <utility>
+#include <vector>
+#include <xmemory>
+#include <xutility>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+
+template <class _FlatMap_Key, class _FlatMap_T, class _FlatMap_Compare = less<_FlatMap_Key>,
+    class _FlatMap_KeyContainer = vector<_FlatMap_Key>, class _FlatMap_MappedContainer = vector<_FlatMap_T>>
+    requires same_as<_FlatMap_Key, typename _FlatMap_KeyContainer::value_type>
+          && same_as<_FlatMap_T, typename _FlatMap_MappedContainer::value_type>
+class flat_map;
+
+struct sorted_unique_t {
+    explicit sorted_unique_t() = default;
+};
+inline constexpr sorted_unique_t sorted_unique{};
+
+template <class _Alloc, class _Key_container, class _Mapped_container>
+concept _Valid_Allocator_for_flat_map =
+    uses_allocator_v<_Key_container, _Alloc> && uses_allocator_v<_Mapped_container, _Alloc>;
+
+template <class _Compare, class _Key_container>
+concept _Valid_Compare_for_container = is_invocable_v<const _Compare&, const typename _Key_container::value_type&,
+    const typename _Key_container::value_type&>;
+
+template <class _Compare>
+concept _Compare_is_transparent = requires { typename _Compare::is_transparent; };
+
+template <class _Key, class _T, class _Key_compare>
+struct _Flat_value_compare {
+    struct value_compare {
+    public:
+        bool operator()(pair<const _Key&, const _T&> _X, pair<const _Key&, const _T&> _Y) const {
+            return _Key_compare_for_val(_X.first, _Y.first);
+        }
+
+        value_compare(_Key_compare _Comp) : _Key_compare_for_val(_Comp) {}
+
+    private:
+        _Key_compare _Key_compare_for_val;
+    };
+};
+
+template <class _Key_container, class _Mapped_container>
+struct _Flat_Container {
+    struct container {
+        _Key_container keys;
+        _Mapped_container values;
+    };
+};
+
+template <class _Ty>
+struct _NODISCARD _Clear_flat_map_scope_guard {

```

Comment:
@achabense Should we put this scope guard in `<xmemory>` and share between both headers?

Reply:
There's already `_Tidy_guard` in `<xmemory>`. I'm not sure whether we should reuse it for flat container adaptors.

Reply:
This utility should be shared finally. However, as both `<flat_map>` and `<flat_set>` haven't been finished, I think it will be ok to keep these guards as-is separately; the sharing-cleanup will be easy to make when either header is merged into the main branch, and can be done together with `sorted_X_t`.